src/scf/newton.jl

```diff
@@ -0,0 +1,420 @@
+
+    # packing routines
+    pack(φ) = pack_arrays(basis, φ)
+    unpack(x) = unpack_arrays(basis, x)
+    packed_proj(δx, x) = pack(proj_tangent(unpack(δx), unpack(x)))
+
+    # mapping of the linear system on the tangent space
+    function f(x)
+        δφ = unpack(x)
+        δφ = proj_tangent(δφ, φproj)
+        Kδφ = apply_K(basis, δφ, φproj, ρproj, occ)
+        Ωδφ = apply_Ω(basis, δφ, φproj, Hproj)
+        pack(Ωδφ .+ Kδφ)
+    end
+
+    # project res on the good tangent space before starting
+    res = proj_tangent(res, φproj)
+
+    # solve (Ω+K) δφ = res on the tangent space with KrylovKit
+    δφ, info = linsolve(f, pack(res);
+                        tol=tol_krylov, verbosity=krylov_verbosity,
+                        # important to specify custom orthogonaliser to keep
+                        # iterations on the tangent space
+                        orth=OrthogonalizeAndProject(packed_proj, pack(φproj)))
+    δφ = unpack(δφ)
+    # ensure proper projection onto the tangent space
+    δφ = proj_tangent(δφ, φ)
+    φ_newton = similar(φ)
+
+    # perform newton_step
+    for ik = 1:Nk
+        φ_newton[ik] = ortho_qr(φ[ik] - δφ[ik])
+    end
+    (φ_newton, δφ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis; ψ0=nothing,
+           tol=1e-6, max_iter=100, φproj=nothing, verbosity=0)
+
+Newton algorithm. Be careful that the starting needs to be not too far from the solution.
+Still contains debugging features.
+If φproj is nothing, we use the successive φ's to define the tangent spaces on which we solve
+Newton's equation. φproj can be useful when we want to fix the tangent space for testing.
+"""
+function newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                tol=1e-6, max_iter=20, φproj=nothing, verbosity=0) where T
+
+    ## setting parameters
+    model = basis.model
+    @assert model.spin_polarization in (:none, :spinless, :collinear)
+    @assert model.temperature == 0 # temperature is not yet supported
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    N = div(div(model.n_electrons, filled_occ), n_spin)
+
+    ## number of kpoints and occupation
+    Nk = length(basis.kpoints)
+    occupation = [filled_occ * ones(T, N) for ik = 1:Nk]
+
+    ## starting point and keep only occupied orbitals
+    if ψ0 === nothing
+        ψ0 = [ortho_qr(randn(Complex{T}, length(G_vectors(kpt)), N))
+              for kpt in basis.kpoints]
+    else
+        for ik in 1:Nk
+            ψ0[ik] = ψ0[ik][:,1:N]
+        end
+    end
+    if φproj != nothing
+        for ik in 1:Nk
+            φproj[ik] = φproj[ik][:,1:N]
+        end
+    end
+
+    ## iterators
+    err = 1
+    k = 0
+
+    ## orbitals, densities and energies to be updated along the iterations
+    φ = deepcopy(ψ0)
+    ρ = compute_density(basis, φ, occupation)
+    prev_energies = nothing
+
+    @printf "\nn     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ\n" "Energy"
+    @printf "---   ---------------   ---------   --------\n"
```

Comment:
Could you refactor the printing into a callback like in the SCF?

Reply:
Yes I can (deep in me I knew I would have to do that at some point :smile: )